adb/adb_protocol.py

```diff
@@ -409,3 +429,128 @@ def StreamingCommand(cls, usb, service, command='', timeout_ms=None):
         timeout_ms=timeout_ms)
     for data in connection.ReadUntilClose():
       yield data.decode('utf8')
+
+  @classmethod
+  def InteractiveShellCommand(cls, conn, cmd=None, strip_cmd=True, delim=None, strip_delim=True, clean_stdout=True):
+    """Retrieves stdout of the current InteractiveShell and sends a shell command if provided
+    TODO: Should we turn this into a yield based function so we can stream all output?
+
+    Args:
+      conn: Instance of AdbConnection
+      cmd: Optional. Command to run on the target.
+      strip_cmd: Optional (default True). Strip command name from stdout.
+      delim: Optional. Delimiter to look for in the output to know when to stop expecting more output
+      (usually the shell prompt)
+      strip_delim: Optional (default True): Strip the provided delimiter from the output
+      clean_stdout: Cleanup the stdout stream of any backspaces and the characters that were deleted by the backspace
+    Returns:
+      The stdout from the shell command.
+    """
+
+    if isinstance(delim, str):
+      delimiter = delim.encode('utf-8')
+
+    # Delimiter may be shell@hammerhead:/ $
+    # The user or directory could change, making the delimiter somthing like root@hammerhead:/data/local/tmp $
+    # Handle a partial delimiter to search on and clean up
+    if delim:
+      user_pos = delim.find(b'@')
+      dir_pos = delim.rfind(b':/')
+      if user_pos != -1 and dir_pos != -1:
+        partial_delim = delim[user_pos:dir_pos+1] # e.g. @hammerhead:
+      else:
+        partial_delim = delim
+    else:
+      partial_delim = None
+
+    stdout = ''
+    stdout_stream = BytesIO()
+    original_cmd = ''
+
+    try:
+
+      if cmd:
+        original_cmd = str(cmd)
+        cmd += '\r'  # Required. Send a carriage return right after the cmd
+        cmd = cmd.encode('utf8')
+
+        # Send the cmd raw
+        bytes_written = conn.Write(cmd)
+
+        if delim:
+          # Expect multiple WRTE cmds until the delim (usually terminal prompt) is detected
+
+          data = b''
+          while partial_delim not in data:
+
+            cmd, data = conn.ReadUntil(b'WRTE')
+            stdout_stream.write(data)
+
+        else:
+          # Otherwise, expect only a single WRTE
+          cmd, data = conn.ReadUntil(b'WRTE')
+
+          # WRTE cmd from device will follow with stdout data
+          stdout_stream.write(data)
+
+      else:
+
+        # No cmd provided means we should just expect a single line from the terminal. Use this sparingly
+        cmd, data = conn.ReadUntil(b'WRTE')
+        if cmd == b'WRTE':
+            # WRTE cmd from device will follow with stdout data
+            stdout_stream.write(data)
+        else:
+            print("Unhandled cmd: {}".format(cmd))
+
+      cleaned_stdout_stream = BytesIO()
+      if clean_stdout:
```

Comment:
you should only have to do this if you're not outputting to a terminal, which will handle backspaces for you. Again, why are you supporting using the interactive shell in a non-interactive space?

Reply:
See my above comments for an explanation